internal/proxy/proxy.go

```diff
@@ -192,165 +192,189 @@ func (p *Proxy) Handle(w http.ResponseWriter, r *http.Request) {
 	}).ServeHTTP(w, r)
 }
 
-func (p *Proxy) handleJenkinsUIRequest(w http.ResponseWriter, r *http.Request, requestLogEntry *log.Entry) (cacheKey string, ns string, noProxy bool) {
+func (p *Proxy) handleJenkinsUIRequest(responseWriter http.ResponseWriter, request *http.Request, requestLogEntry *log.Entry) (cacheKey string, namespace string, noProxy bool) {
 	//redirect determines if we need to redirect to auth service
 	needsAuth := true
 	noProxy = true
 
-	//redirectURL is used for auth service as a target of successful auth redirect
-	redirectURL, err := url.ParseRequestURI(fmt.Sprintf("%s%s", strings.TrimRight(p.redirect, "/"), r.URL.Path))
+	// redirectURL is used for auth service as a target of successful auth redirect
+	redirectURL, err := url.ParseRequestURI(fmt.Sprintf("%s%s", strings.TrimRight(p.redirect, "/"), request.URL.Path))
 	if err != nil {
-		p.HandleError(w, err, requestLogEntry)
+		p.HandleError(responseWriter, err, requestLogEntry)
 		return
 	}
 
 	//If the user provides OSO token, we can directly proxy
-	if _, ok := r.Header["Authorization"]; ok { //FIXME Do we need this?
+	if _, ok := request.Header["Authorization"]; ok { //FIXME Do we need this?
 		needsAuth = false
 	}
 
-	if tj, ok := r.URL.Query()["token_json"]; ok { //If there is token_json in query, process it, find user info and login to Jenkins
-		if len(tj) < 1 {
-			p.HandleError(w, errors.New("could not read JWT token from URL"), requestLogEntry)
+	namespace, noProxy = p.processTokenJSON(responseWriter, request, requestLogEntry, redirectURL)
+
+	cacheKey, namespace, noProxy, needsAuth = p.checkCookies(responseWriter, request, requestLogEntry)
+
+	//Check if we need to redirect to auth service
+	if needsAuth {
+		redirAuth := GetAuthURI(p.authURL, redirectURL.String())
+		requestLogEntry.Infof("Redirecting to auth: %s", redirAuth)
+		http.Redirect(responseWriter, request, redirAuth, 301)
+	}
+	return
+}
+
+// processTokenJSON processes token_json if present in query and find user info and login to Jenkins
+func (p *Proxy) processTokenJSON(responseWriter http.ResponseWriter, request *http.Request, requestLogEntry *log.Entry, redirectURL *url.URL) (namespace string, noProxy bool) {
+
+	// Checks if token_json is present in the query
+	if tokenJSON, ok := request.URL.Query()["token_json"]; ok {
```

Comment:
It might make sense to extract this into something like `func isAuthenticatedRequest(request *http.Request) bool`. If you then pull up the if statement again, the method name documents the alogrithm.

```golang
func (p *Proxy) handleJenkinsUIRequest(responseWriter http.ResponseWriter, request *http.Request, requestLogEntry *log.Entry) (cacheKey string, namespace string, noProxy bool) {
        if isAuthenticatedRequest(request) {
	   namespace, noProxy = p.processAuthenticatedRequest(responseWriter, request, requestLogEntry, redirectURL)
        } else {
           return redirectToAuth(...)
       }
 
        if isWaitLoopRequest(request) {
	    cacheKey, namespace, noProxy, needsAuth = p.checkCookies(responseWriter, request, requestLogEntry)
         }
}
```

Reply:
Not sure which code segment could be kept in isWaitLoopRequest(request)